File: 0179_largest_number.py
# https://leetcode-cn.com/problems/largest-number/

class Solution:
    def largestNumber(self, nums):
        """
        Inputs:
            nums: list[int]
        Outpus:
            res: str
        """

        n = len(nums)

        # Sort.
        # If 'ab' < 'ba', then b should be in front of a.
        for i in range(n-1):
            for j in range(i+1, n):
                if str(nums[i]) + str(nums[j]) < str(nums[j]) + str(nums[i]):
                    nums[i], nums[j] = nums[j], nums[i]

        res = ''.join([str(i) for i in nums])

        # Check if the final result is '0'.
        if int(res) == 0:
            res = '0'

        return res

File: Cpp/0179_largest_number.cpp
// https://leetcode.cn/problems/largest-number/

class Solution {
public:
    string largestNumber(vector<int>& nums) {
        vector<string> nums_s;
        for (int num: nums) {
            string num_s = to_string(num);
            nums_s.push_back(num_s);
        }

        sort(nums_s.begin(), nums_s.end(),
            [](string s1, string s2) {
                string s1s2 = s1 + s2;
                string s2s1 = s2 + s1;
                return s1s2 > s2s1;
            });

        string res = "";
        for (string num_s: nums_s) {
            res = res + num_s;
        }
        if (res[0] == '0') {
            res = "0";
        }
        return res;
    }
};
